pkg/api/defaults.go

```diff
@@ -792,46 +792,51 @@ func (cs *ContainerService) setWindowsProfileDefaults(isUpgrade, isScale bool) {
 			windowsProfile.SSHEnabled = to.BoolPtr(DefaultWindowsSSHEnabled)
 		}
 
+		// Default to aks-engine WIndows Server 2019 docker image
+		defaultImageConfig := AKSWindowsServer2019OSImageConfig
+		if windowsProfile.WindowsPublisher == WindowsServer2019OSImageConfig.ImagePublisher && windowsProfile.WindowsOffer == WindowsServer2019OSImageConfig.ImageOffer {
+			// Use 'vanilla' Windows Server 2019 images
+			defaultImageConfig = WindowsServer2019OSImageConfig
+		} else if cs.Properties.OrchestratorProfile.KubernetesConfig.NeedsContainerd() {
+			// Use to using aks-engine Windows Server 2019 conatinerD VHD
+			defaultImageConfig = AKSWindowsServer2019ContainerDOSImageConfig
+		}
+
 		// This allows caller to use the latest ImageVersion and WindowsSku for adding a new Windows pool to an existing cluster.
 		// We must assure that same WindowsPublisher and WindowsOffer are used in an existing cluster.
-		if windowsProfile.WindowsPublisher == AKSWindowsServer2019OSImageConfig.ImagePublisher && windowsProfile.WindowsOffer == AKSWindowsServer2019OSImageConfig.ImageOffer {
+		if windowsProfile.WindowsPublisher == defaultImageConfig.ImagePublisher && windowsProfile.WindowsOffer == defaultImageConfig.ImageOffer {
 			if windowsProfile.WindowsSku == "" {
-				windowsProfile.WindowsSku = AKSWindowsServer2019OSImageConfig.ImageSku
+				windowsProfile.WindowsSku = defaultImageConfig.ImageSku
 			}
 			if windowsProfile.ImageVersion == "" {
-				if windowsProfile.WindowsSku == AKSWindowsServer2019OSImageConfig.ImageSku {
-					windowsProfile.ImageVersion = AKSWindowsServer2019OSImageConfig.ImageVersion
-				} else {
-					windowsProfile.ImageVersion = "latest"
-				}
-			}
-		} else if windowsProfile.WindowsPublisher == WindowsServer2019OSImageConfig.ImagePublisher && windowsProfile.WindowsOffer == WindowsServer2019OSImageConfig.ImageOffer {
-			if windowsProfile.WindowsSku == "" {
-				windowsProfile.WindowsSku = WindowsServer2019OSImageConfig.ImageSku
-			}
-			if windowsProfile.ImageVersion == "" {
-				if windowsProfile.WindowsSku == WindowsServer2019OSImageConfig.ImageSku {
-					windowsProfile.ImageVersion = WindowsServer2019OSImageConfig.ImageVersion
+				if windowsProfile.WindowsSku == defaultImageConfig.ImageSku {
+					windowsProfile.ImageVersion = defaultImageConfig.ImageVersion
 				} else {
 					windowsProfile.ImageVersion = "latest"
 				}
 			}
 		} else {
 			if windowsProfile.WindowsPublisher == "" {
-				windowsProfile.WindowsPublisher = AKSWindowsServer2019OSImageConfig.ImagePublisher
+				windowsProfile.WindowsPublisher = defaultImageConfig.ImagePublisher
 			}
 			if windowsProfile.WindowsOffer == "" {
-				windowsProfile.WindowsOffer = AKSWindowsServer2019OSImageConfig.ImageOffer
+				windowsProfile.WindowsOffer = defaultImageConfig.ImageOffer
```

Comment:
this seems like it could get into a bad state if the profile for the windows publisher doesn't not equal defaultimageconfig.   Would it be simpler to calculate the default config based on profile (which seems to be happening partially above) and then update accordingly?

Reply:
If the user is using an offer we don't curate here (MicrosoftWindowsServer 2004 images for example) we don't know what values to use here.
Maybe it would be better to turn this into an else-if, verify the publisher/offer match defaultImageConfig, before setting them and log and error if they don't match?
Thoughts?

Reply:
There are many different offers available to users that we do not curate in aks-engine (1909, 2004, 20H2 images, etc).
I don't think it is possible for us to maintain default values for all of these offers.

I agree users can get into a bad state if they only specify windowsPublisher but am not sure what else to do here.
If things get messed up ARM does have a detailed message saying the publisher/offer/sku/image cannot be found.

Reply:
Maybe we can nest these and only set offer if windowsPublisher == defaultImageConfig.ImagePublisher?

Reply:
Makes sense on all the versions.  If we assume they can get into a bad state could this logic be simplified and spit out a warning or maybe even an error that that tells they are going to get into a bad state with the values provided if they don't use one of our known defaults?

Reply:
> Maybe we can nest these and only set offer if windowsPublisher == defaultImageConfig.ImagePublisher?

for some reason this last comment didn't show up in my ui but I think we were thinking in the same direction

Reply:
Good idea. I'll check if we think things are not good, use the default image, a log a warming.